File: BinaryTree/222.cpp
#include <iostream>
#include <vector>


using namespace std;
// Definition for a binary tree node.
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right)
        : val(x), left(left), right(right) {}
};
class Solution
{
public:
    int countNodes(TreeNode* root) {
        if (root = 0)
            return NULL;
        TreeNode* left = root->left;
        TreeNode* right = root->right;
        int leftLength = 0;
        int rightLength = 0;
        while (left)
        {
            left = left->left;
            leftLength++;
        }
        while (right)
        {
            right = right->right;
            rightLength++;
        }
        if (leftLength == rightLength) {
            return (2 << leftLength) - 1;
        }
        return coutnNodes(left) + countNodes(right) + 1;
    }
};

File: BinaryTree/222.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def countNodes(self, root: Optional[TreeNode]) -> int:
        if root == None:
            return 0
        left_length = 0
        right_length = 0
        left = root.left
        right = root.right
        while left:
            left = left.left
            left_length += 1
        while right:
            right = right.right
            right_length += 1
        if left_length == right_length:
            return 2 ** (left_length + 1) - 1
        return self.countNodes(root.left) + self.countNodes(root.right) + 1
